### PR TITLE
Add tree sequence metadata and metadata_schema to C API

### PR DESCRIPTION
**Pull request by benjeffery**
C API part of #534. Adds `metadata` and `metadata_schema` fields to table collection, with accessors on tree sequence. These store arbitrary bytes and are optional in the file format.

### CODE REVIEW

File: c/tskit/tables.c [hunk 1]
```diff
@@ -7000,6 +7051,44 @@ tsk_table_collection_read_format_data(tsk_table_collection_t *self, kastore_t *s
...
     self->file_uuid[TSK_UUID_SIZE] = '\0';
+
+    ret = kastore_containss(store, "metadata");
+    if (ret != 1 && ret != 0) {
+        ret = tsk_set_kas_error(ret);
+        goto out;
+    }
+    if (ret == 1) {
+        ret = kastore_gets_int8(
+            store, "metadata", (int8_t **) &metadata, (size_t *) &metadata_length);
```
Comment (benjeffery) [inline c/tskit/tables.c:7062]: The cast here and on line 7079 are effectively what is happening in the table metadata already, but I think it isn't _strictly_ right. It is practically fine as it works in our tests, but it's my understanding that C99 only asks that `CHAR_BIT => 8` not `CHAR_BIT==8`

**Comment by benjeffery** [comment]
Looks like I could get a little more coverage here, setting back to draft.

**Comment by benjeffery** [comment]
Added a test on the TS methods. Looks like the rest of the coverage misses are hard error paths. Ready for review.

File: c/tskit/tables.c (hunk 1 above)
Comment (jeromekelleher) [reply c/tskit/tables.c:7062]: Yeah, that is a potential issue. Probably we should just put in a ``assert(CHAR_BIT == 8)`` in the code somewhere central. It's not worth getting worried about more than that though.

File: c/tskit/tables.c [hunk 2]
```diff
@@ -7000,6 +7051,45 @@ tsk_table_collection_read_format_data(tsk_table_collection_t *self, kastore_t *s
     }
     memcpy(self->file_uuid, uuid, TSK_UUID_SIZE);
     self->file_uuid[TSK_UUID_SIZE] = '\0';
+
+    ret = kastore_containss(store, "metadata");
```
Comment (jeromekelleher) [inline c/tskit/tables.c:7055]: Good catch, this was potentially masking errors. I'd say using ``if (ret < 0) `` is *slightly* more idiomatic for detecting the error condition, but it's not worth holding the PR up for.

File: python/tests/test_file_format.py [hunk 3]
```diff
@@ -825,8 +827,9 @@ def verify_missing_fields(self, ts):
         with kastore.load(self.temp_file) as store:
             all_data = dict(store)
         for key in all_data.keys():
-            # We skip this key as it is optional
-            if "metadata_schema" not in key:
+            # We skip these keys as it is optional
+            if "metadata_schema" not in key and key != "metadata":
+                print(key)
```
Comment (jeromekelleher) [inline python/tests/test_file_format.py:832]:
```suggestion
```

Comment (benjeffery) [reply python/tests/test_file_format.py]: Facepalm! So close to a "hole in one" PR. Ah well, will try again next time!

File: c/tskit/tables.c (hunk 2 above)
Comment (benjeffery) [reply c/tskit/tables.c:7055]: Agreed, should be that.

**Comment by benjeffery** [comment]
@jeromekelleher did you see my comment above about the casting to `char`?

**Comment by benjeffery** [comment]
Oops just seen the reply, ignore!

**Comment by jeromekelleher** [comment]
I seem to have messed the PR up here  - I thought you were off today @benjeffery so I force pushed a couple of changes. Looks like I broke stuff.

Crap, I think I pulled in kastore updates by ``git commit -a``.

**Comment by jeromekelleher** [comment]
I made the mess, I'll fix it up!

**Comment by benjeffery** [comment]
I'm off, but can quickly push the branch back to how it was?

**Comment by jeromekelleher** [comment]
> I'm off, but can quickly push the branch back to how it was?

Na, thanks, it's an easy fix.

File: c/tskit/tables.c (hunk 1 above)
Comment (jeromekelleher) [reply c/tskit/tables.c:7062]: opened isssue to track, #645.

**Pull request merged**